Static type support for rosidl_typesupport_c

pablogs9 opened the pull request:
Hello @sloretz and @clalancette,

here at micro-ROS, we are wondering if it is possible to implement the changes proposed here to allow `rosidl_typesupport_c` to have some kind of statically linked typesupport dispatch for those environments where dynamic library are not allowed (embedded).

Can we discuss this feature here? What are your opinions? 

CC: @ralph-lange @JanStaschulat

pablogs9 committed:
Static type support initial

Signed-off-by: Your Name <you@example.com>

Default to OFF

Signed-off-by: Your Name <you@example.com>

Linter

Signed-off-by: Your Name <you@example.com>

Fix

Signed-off-by: Your Name <you@example.com>

Refactor

Reafactor

Signed-off-by: Your Name <you@example.com>

Lint

Signed-off-by: Your Name <you@example.com>

clalancette reviewed (commented):
In general, I think this would be a fine feature to have.

I do have some concerns about the implementation.  In no particular order:
1. The `#ifdef` in type_support_dispatch.hpp looks pretty ugly.  It seems to me that it would be better to split the code underneath the `#ifdef` into a separate file (`dynamic_support_dispatch.hpp` or something), and then just call out to it as appropriate.
2. The new flag to CMakeLists.txt (`ROSIDL_TYPESUPPORT_STATIC_TYPESUPPORT`) look somewhat redundant with the `${rosidl_typesupport_c_LIBRARY_TYPE}" STREQUAL "STATIC"` code.  Is there any way to combine that logic together?
3. Similarly, the logic in the empy files feels like it would apply in both the old "STATIC" case and the new "ROSIDL_TYPESUPPORT_STATIC_TYPESUPPORT" case, but I may be misunderstanding something here.

@sloretz any thoughts here?

pablogs9 committed:
Separate dynamic handling

Signed-off-by: Your Name <you@example.com>

Inferring flags from build type

Signed-off-by: Your Name <you@example.com>

Remove extra include

Signed-off-by: Your Name <you@example.com>

Update

Fix rebase

Signed-off-by: Your Name <you@example.com>

Fixes

Add messages

Update

Fix

Fix

Messages

Update

Linter

Lint

Remove messages

Signed-off-by: Your Name <you@example.com>

pablogs9 commented:
Hello @clalancette, thanks a lot for your comments. I'm answering here:

1. I have created `dynamic_support_dispatch.hpp` with the logic associated to the dynamic library handling
2. I have found that by default, at least in the tests, `BUILD_SHARED_LIBS` is not set, so [this line](https://github.com/ros2/rosidl_typesupport/blob/3bd7d816e62ec05390af09a761ea3e0d7bb7b9a6/rosidl_typesupport_c/CMakeLists.txt#L133) is reached and probably the `rosidl_typesupport_c` library is being built as a static one. At some point, the build type (static or dynamic) of the `rosidl_typesupport_c` does not determine if the actual typesupport is being handled statically or dynamically. If fact a static library can use `rcutils` functions to search and use the dynamic library of the typesupport for a message. 
So, although I have modified it to use this flags, I wonder if we should keed the original `option(ROSIDL_TYPESUPPORT_STATIC_TYPESUPPORT "Enable static typesupport" OFF)`
3. Sorry I don't understand which are the "empy files". Could you explain this comment?

pablogs9 committed:
Allow single typessuport

Signed-off-by: Your Name <you@example.com>

pablogs9 committed:
Merge branch 'master' into feature/static_typesupport

bjsowa commented:
@clalancette Could you please recheck this PR?
